core/src/main/java/org/apache/accumulo/core/fate/FateExecutor.java

```diff
@@ -0,0 +1,477 @@
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus.FAILED;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus.FAILED_IN_PROGRESS;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus.IN_PROGRESS;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus.SUBMITTED;
+import static org.apache.accumulo.core.fate.ReadOnlyFateStore.TStatus.SUCCESSFUL;
+import static org.apache.accumulo.core.util.ShutdownUtil.isIOException;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TransferQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import org.apache.accumulo.core.clientImpl.AcceptableThriftTableOperationException;
+import org.apache.accumulo.core.fate.Fate.TxInfo;
+import org.apache.accumulo.core.fate.FateStore.FateTxStore;
+import org.apache.accumulo.core.util.ShutdownUtil;
+import org.apache.accumulo.core.util.Timer;
+import org.apache.accumulo.core.util.threads.ThreadPoolNames;
+import org.apache.accumulo.core.util.threads.ThreadPools;
+import org.apache.accumulo.core.util.threads.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * Handles finding and working on FATE work. Only finds/works on fate operations that it is assigned
+ * to work on defined by 'fateOps'
+ */
+public class FateExecutor<T> {
+  private static final Logger log = LoggerFactory.getLogger(FateExecutor.class);
+  private final Logger runnerLog = LoggerFactory.getLogger(TransactionRunner.class);
+
+  private final T environment;
+  private final Fate<T> fate;
+  private final Thread workFinder;
+  private final TransferQueue<FateId> workQueue;
+  private final String poolName;
+  private final ThreadPoolExecutor transactionExecutor;
+  private final Set<TransactionRunner> runningTxRunners;
+  private final Set<Fate.FateOperation> fateOps;
+  private final ConcurrentLinkedQueue<Integer> idleCountHistory = new ConcurrentLinkedQueue<>();
+
+  public FateExecutor(Fate<T> fate, T environment, Set<Fate.FateOperation> fateOps, int poolSize) {
+    final String operatesOn = fate.getStore().type().name().toLowerCase() + "."
+        + fateOps.stream().map(fo -> fo.name().toLowerCase()).collect(Collectors.joining("."));
+    final String transactionRunnerPoolName =
+        ThreadPoolNames.MANAGER_FATE_POOL_PREFIX.poolName + operatesOn;
+    final String workFinderThreadName = "fate.work.finder." + operatesOn;
```

Comment:
Example pool name: `accumulo.pool.manager.fate.meta.table_split.system_split.table_delete_range.table_merge`
This is long, but descriptive